javalib/src/main/scala/scala/scalanative/nio/fs/NativePosixFileAttributeView.scala

```diff
@@ -47,77 +48,85 @@ final class NativePosixFileAttributeView(path: Path, options: Array[LinkOption])
         case (flag, value) => if (perms.contains(value)) mask = mask | flag
       }
       if (stat.chmod(toCString(path.toString), mask) != 0) {
-        throw new IOException()
+        throwIOException()
       }
     }
 
-  override def getOwner(): UserPrincipal =
-    attributes.owner
+  override def getOwner(): UserPrincipal = attributes.owner
 
   override def setGroup(group: GroupPrincipal): Unit =
     Zone { implicit z =>
       val _group = getGroup(toCString(group.getName))
       val err    = unistd.chown(toCString(path.toString), -1.toUInt, !(_group._2))
 
       if (err != 0) {
-        throw new IOException()
+        throwIOException()
       }
     }
 
-  override def readAttributes(): BasicFileAttributes =
-    attributes
+  override def readAttributes(): BasicFileAttributes = attributes
 
-  private lazy val attributes =
+  private def attributes =
     new PosixFileAttributes {
-      private def fileStat()(implicit z: Zone) =
-        getStat()
-
-      private def fileMode()(implicit z: Zone) =
-        !(fileStat()._13)
+      private[this] var st_dev: stat.dev_t         = _
+      private[this] var st_rdev: stat.dev_t        = _
+      private[this] var st_ino: stat.ino_t         = _
+      private[this] var st_uid: stat.uid_t         = _
+      private[this] var st_gid: stat.gid_t         = _
+      private[this] var st_size: unistd.off_t      = _
+      private[this] var st_atime: time.time_t      = _
+      private[this] var st_mtime: time.time_t      = _
+      private[this] var st_ctime: time.time_t      = _
+      private[this] var st_blocks: stat.blkcnt_t   = _
+      private[this] var st_blksize: stat.blksize_t = _
+      private[this] var st_nlink: stat.nlink_t     = _
+      private[this] var st_mode: stat.mode_t       = _
```

Comment:
Do these cached values play well with file attributes changing? i.e. what happens if file changes after the attributes where read and then the attribute info is requested later?

Reply:
The updated semantics should be the same as on the jvm. It isn't clear from the javadoc, but when you call `Files.readAttributes`, you get a snapshot of the file attributes at the moment that the sys call to stat is invoked. For example, I ran this in the console:
```
scala> import java.nio.file._
import java.nio.file._

scala> val dir = Files.createTempDirectory("foo")
dir: java.nio.file.Path = /var/folders/7q/s3vq91lx3zv9fw4l5l5c43nr0000gn/T/foo5386655488881479548

scala> val dirAttrs = Files.readAttributes(dir, classOf[attribute.BasicFileAttributes])
dirAttrs: java.nio.file.attribute.BasicFileAttributes = sun.nio.fs.UnixFileAttributes$UnixAsBasicFileAttributes@1856eae9

scala> Files.delete(dir)

scala> Files.createFile(dir)
res1: java.nio.file.Path = /var/folders/7q/s3vq91lx3zv9fw4l5l5c43nr0000gn/T/foo5386655488881479548

scala> val attrs = Files.readAttributes(dir, classOf[attribute.BasicFileAttributes])
attrs: java.nio.file.attribute.BasicFileAttributes = sun.nio.fs.UnixFileAttributes$UnixAsBasicFileAttributes@285bb22a

scala> attrs.isRegularFile
res2: Boolean = true

scala> attrs.isDirectory
res3: Boolean = false

scala> dirAttrs.isRegularFile
res4: Boolean = false

scala> dirAttrs.isDirectory
res5: Boolean = true
```

Reply:
Good point, it's all good then.